staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/apiserver.go

```diff
@@ -244,6 +248,55 @@ func (c completedConfig) New(delegationTarget genericapiserver.DelegationTarget)
 		go finalizingController.Run(5, context.StopCh)
 		return nil
 	})
+	s.GenericAPIServer.AddPostStartHookOrDie("crd-discovery-available", func(context genericapiserver.PostStartHookContext) error {
+		return wait.PollImmediateUntil(100*time.Millisecond, func() (bool, error) {
+			// only check if we have a valid list for a given resourceversion
+			if !s.Informers.Apiextensions().V1().CustomResourceDefinitions().Informer().HasSynced() {
+				return false, nil
+			}
+
+			// The returned group and resource lists might be non-nil with partial results even in the
+			// case of non-nil error.  If API aggregation fails, we still want our other discovery information because the CRDs
+			// may all be present.
+			_, serverGroupsAndResources, discoveryErr := crdClient.Discovery().ServerGroupsAndResources()
+			if discoveryErr != nil {
+				klog.V(2).Info(discoveryErr)
+			}
+
+			serverCRDs, err := s.Informers.Apiextensions().V1().CustomResourceDefinitions().Lister().List(labels.Everything())
```

Comment:
rather than relisting all CRDs every time through the loop, I would have expected to do an initial list to find served+established CRDs that existed at start to figure out what resources we are waiting for, then loop until those resources exist in discovery (or that CRD version no longer exists as served+established)